### PR TITLE
[US:1] Giant Huge Mega Update

### PR DESCRIPTION
**Pull request by JonMcDevitt**
Added ClassDay and AvailableCoursesTable (database) classes and corresponding
tests.

Refactored basically all of the classes to the user interface (ui) directory.
Also added several classes to the backend directory.

### CODE REVIEW

**Comment by nbroderick42** [comment]
Awaiting resolution of merge conflicts at b196c49

File: src/main/java/com/project/backend/AvailableCoursesTable.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            int index = binarySearch(c);
+            if(index >= 0) {
+                System.err.println("Cannot add to table, course already present.");
+            } else {
+                courses.add(-index-1, c);
+            }
+        }
+    }
+
+    private int binarySearch(Course c) {
```
Comment (nbroderick42) [inline src/main/java/com/project/backend/AvailableCoursesTable.java]: The Collections has a `binarySearch` implementation that will work here. Seems a better choice than rolling our own binary search. Alternately, a `Map` may be a better choice here for the backing data structure if we expect reading to be the most common operation.

File: src/main/java/com/project/backend/AvailableCoursesTable.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        } else {
+            return -mid-2;
+        }
+    }
+
+    static AvailableCoursesTable getInstance() {
+        return instance;
+    }
+
+    private static void populate() {
```
Comment (nbroderick42) [inline src/main/java/com/project/backend/AvailableCoursesTable.java]: This seems to duplicate the work done in the `private` constructor.

File: src/main/java/com/project/backend/ClassDay.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package com.project.backend;
+
+import java.util.ArrayList;
+import java.util.Date;
```
Comment (nbroderick42) [inline src/main/java/com/project/backend/ClassDay.java]: The `Date` implementation in Java is a bit... broken. If possible we should probably find an alternative for this - probably the `java.time` package and its classes/methods.

See http://stackoverflow.com/questions/1969442/whats-wrong-with-java-date-time-api for some rationale.

File: src/main/java/com/project/backend/AvailableCoursesTable.java (hunk 2 above)
Comment (JonMcDevitt) [inline src/main/java/com/project/backend/AvailableCoursesTable.java]: Yes, but it won't let me call functions like populate from within the constructor.

Comment (nbroderick42) [inline src/main/java/com/project/backend/AvailableCoursesTable.java]: Looking at it more closely, it looks like `populate` should a member method, not a class method. Try that?

**Pull request closed without merging**